### DISCUSSION TITLE
Brighter Feedback

### DISCUSSION THREAD

**Discussion by iancooper** [BrighterCommand/Brighter#2811] [General]
@iancooper please could you help us validate more issues we're anticipating in situations with:
- Number of Brighter instances running reaching ~50
- Rate of messages being produced upwards of 250 per second

### Our current understanding
During `PostAsync` the following happens:
- Message is deposited into the outbox
- Message is attempted to be cleared from the outbox
- Also `CheckOutstandingMessages`

and during the sweeper process:
- Get all outstanding messages (in our case with Kafka by default at least 5 seconds old)
- Attempt to clear all of them from the outbox

### Potential Issue
Race condition we're concerned with are:
- Tick 0: Brighter process 1 begins `PostAsync` on message `A`
- Tick 1 to 50 million: Brighter process 1 still awaiting success response for message `A`
- Tick 50 million and 1 (5s later): Brighter process 2 begins sweeping and picks up message `A` as outstanding
- ...: Message `A` is published twice

Since an attempt to clear the outbox will be made by both sweepers which each Brighter process will have and as part of every `PostAsync` this would lead to a number of duplicate messages being published proportionate to the rate of messages being published and the number of sweeper processes competing against each other.

### Potential Solution
Assuming our understanding is correct, we're thinking it may make sense to implement a kind of 'visibility timeout'? So during `DynamoDbOutbox.OutstandingMessages` we would query for:
```
KeyExpression => TopicShard = :v_TopicShard and CreatedTime < :v_CreatedTime
FilterExpression => VisibilityTimeout > :v_VisibilityTimeout
```

Where we also update the visibility timeout again inside of `DynamoDbOutbox.OutstandingMessages` to be `now + {ConfiguredVisibilityTimeout}`.

Wouldn't be perfect as there's no consistent reads for the GSI but would help mitigate those race conditions.

What do you think?

**Comment by iancooper** [comment]
For the time being, is it possible and would it make sense to effectively 'turn off' any sort of outbox with Brighter in it's current state? We understand we'd potentially be trading off throughput but gaining our preference to not unknowingly drop messages.

from @jtsalva moved from https://github.com/BrighterCommand/Brighter/issues/2810

**Comment by iancooper** [comment]
> During PostAsync the following happens:
> 
>     Message is deposited into the outbox
>     Message is attempted to be cleared from the outbox
>     Also CheckOutstandingMessages
> 

Sending a message in Brighter involves two steps:

- **DepositPost**: Add your outgoing message to the outbox (in memory if you have not supplied one), participating in a transaction if you use configure a transaction provider for use with the outbox that you DI into your handler. Returns a list of added messages.
- **ClearAsync**: Dispatches a list of messages using your configured messaging producer (and message mappers).

PostAsync does DepositPost followed by ClearAsync on the list of saved messages, and does not participate in your transaction. 

> and during the sweeper process:
> 
>     Get all outstanding messages (in our case with Kafka by default at least 5 seconds old)
>     Attempt to clear all of them from the outbox

Yes


> Race condition we're concerned with are:
> 
>     Tick 0: Brighter process 1 begins PostAsync on message A
>     Tick 1 to 50 million: Brighter process 1 still awaiting success response for message A
>     Tick 50 million and 1 (5s later): Brighter process 2 begins sweeping and picks up message A as outstanding
>     ...: Message A is published twice

Yep. In distributed systems in order for you to have guaranteed delivery, you must accept the possibility of duplicates. This is why we refer to _guaranteed, at least once_. The only way to get guaranteed, exactly once is to de-duplicate at the consumer side. Kafka has some support for that, but we also offer an [Inbox](https://brightercommand.gitbook.io/paramore-brighter-documentation/guaranteed-at-least-once/brighterinboxsupport) which will allow you to reject a message that you have already successfully handled.

This is a known limitation of distributed systems.

>Assuming our understanding is correct, we're thinking it may make sense to implement a kind of 'visibility timeout'? 
So you could do this, to effectively limit how long before you forward a failed message. It might have some utility outside of the duplicate use case i.e. at a certain point this message would have no meaning. If it is just duplication, I would tend to rely on an Inbox, because the race condition is usually rare, but we could raise a request for a feature that says: outbox messages time out.

@preardon Any thoughts on this idea? Something you have seen?

**Reply by preardon** [reply]
One question I would be asking is around Bulk (This is what I am using for bigger loads)

If bulk is not enabled and used in high load scenarios this can significantly increase dispatch times, I can commit and dispatch thousands of messages per second with bulk, without it was a very low number.

Bulk outbox support : https://brightercommand.gitbook.io/paramore-brighter-documentation/guaranteed-at-least-once/brighteroutboxsupport#bulk-deposit

It looks like we only have bulk dispatch plumbed in for Async Sweeper.

one thing I would ask, is do you need the message sent immediately, I tend to use Implicit sending for the majority of my messages (Use Deposit, and let the sweeper do the dispatch) this leads to more performant requests (if they are API, as the application thread no longer needs to worry about dispatching the message, also it allows you to have your messages take part in a transaction

**Reply by jtsalva** [reply]
For our specific case:
- the messages are time-sensitive and should be sent with minimal delay
- we do not need any support for transactions
- it's not an API, so no concerns with request performance

We've not considered bulk so will look into that more, thank you! Curious what sorts of intervals you've found works well for your sweeper?

**Reply by jtsalva** [reply]
How did you avoid throttling with thousands of messages per second? Even with bulk I would've expected that to exceed DynamoDb limits for a single partition

**Reply by iancooper** [reply]
>How did you avoid throttling with thousands of messages per second? Even with bulk I would've expected that to exceed >DynamoDb limits for a single partition

@jtsalva I suspect because @preardon would be on Azure and not using DynamoDb. So the DynamoDb issues are an AWS problem

**Reply by preardon** [reply]
@jtsalva Ian is correct, I am SQL Server and Azure Service Bus.

If you have time sensitive then explicit send is important I agree, the thing to note (And reason We added bulk) is that at the minute, without bulk it works this way

Post Async
- Deposit Async
  -  Add Messages
- Clear Messages
  - For each Message
    -  Get From Outbox
    -  Dispatch to Broker
    -  Mark Dispatched

that being said I've just had a look and it doesn't look to me that Dynamo supports Bulk Depositing, also it looks like we only support bulk dispatching inside of the sweeper

@iancooper  we should chat about if we want bulk dispatching outside of the sweeper, from what I can see it does not require a bulk outbox just a bulk transport

@jtsalva in your scenario are you only doing Post for single messages at a time? also do you have any telemetry

**Reply by jtsalva** [reply]
Currently, yes we're doing PostAsync for each and every message. 

Are there any metrics you're particularly interested in?

Also to note, we're currently in production only with the in-memory outbox at the moment due to throttling issues being addressed in #2810

**Reply by jtsalva** [reply]
Not sure how bulk depositing would work for our application, essentially we're just mapping messages from one Kafka topic to another Kafka topic. We're using Brighter for the message consuming/handling. It's our understanding that the underlying Confluent Kafka SDK [can only get one message at a time](https://github.com/confluentinc/confluent-kafka-dotnet/issues/1451) unlike in Java land for whatever reason, although we're aware in Brighter you _can_ configure to have batches.

**Reply by jtsalva** [reply]
Regarding duplicates, I agree it makes sense it's accepted behaviour as part of a distributed system, was just wondering if there _could_ be work done to reduce chances as a 'best effort'. But now I know we can tell our consumers to implement an inbox if needed.

Main concern was during high load because we do call PostAsync for each message there would be the check for outstanding messages every time which would contain all messages within the Kafka ack response time period. But I've no data around that so only one way to find out!

**Reply by preardon** [reply]
@jtsalva excellent, this makes a lot of sense now, thank you for the context. There are no "Application state changes" so transactions are not a concern, and you are only processing one Message at a time so bulk is not a concern.  One thing you could explore is move to just using Deposit, and set your sweeper minimum age to 0

**Reply by iancooper** [reply]
>But now I know we can tell our consumers to implement an inbox if needed.
Because we have an Inbox I have no looked to deeply into it but you can also configure the Kafka consumer to de-duplicate for you: https://docs.confluent.io/kafka/design/delivery-semantics.html which Confluent claim offers exactly once. However, I suspect that because their transactional scope does not encompass any transactional scope that you have, this doesn't really help if you want to use an outbox

**Reply by iancooper** [reply]
>if we want bulk dispatching outside of the sweeper
Let's pick this up at the next Brighter meeting

**Reply by iancooper** [reply]
@preardon Any thoughts on a visibility timeout or TTL for a message inside the outbox?

**Reply by jtsalva** [reply]
Reading more into Kafka's de-duplication 'The idempotent delivery option guarantees that resending a message will not result in duplicate entries in the log, and that log order is maintained. To achieve this, the broker assigns each producer an ID and deduplicates messages using a sequence number that is sent by the producer with every message.' - my understanding is that wouldn't help in the (probably rare) case where multiple sweepers pick up the same set of messages at the same time from different Brighter instances as they'll have different producer IDs

**Reply by jtsalva** [reply]
Oh I see what you mean now by 'their transactional scope does not encompass any transactional scope that you have' yep

**Reply by iancooper** [reply]
>where multiple sweepers pick up the same set of messages 
Use just the one sweeper, and restart on crash

**Reply by jtsalva** [reply]
> Use just the one sweeper, and restart on crash

How can we ensure only 1 sweeper is running out of the many Brighter processes?

**Reply by iancooper** [reply]
>How can we ensure only 1 sweeper is running out of the many Brighter processes?
An in-process sweeper only uses an in-memory buffer, so you can use many of them.

An out-of-process sweeper, we recommend that you only have one that you deploy as a separate item. There is an example with that.

**Comment by iancooper** [comment]
@jtsalva ^^^
